src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -10,3 +10,297 @@ pub mod ops;
+    }
+
+    fn build_name_constraints_subtrees(
+        &self,
+        subtrees: SequenceOfSubtrees<'work>,
+    ) -> Vec<GeneralName<'work>> {
+        subtrees.unwrap_read().clone().map(|x| x.base).collect()
+    }
+
+    fn build_name_constraints(
+        &self,
+        constraints: &mut AccumulatedNameConstraints<'work>,
+        working_cert: &'a Certificate<'work>,
+    ) -> Result<(), ValidationError> {
+        let extensions: Extensions<'work> = working_cert
+            .extensions()
+            .map_err(|e| ValidationError::Policy(PolicyError::DuplicateExtension(e)))?;
+        if let Some(nc) = extensions.get_extension(&NAME_CONSTRAINTS_OID) {
+            let nc: NameConstraints<'work> = nc.value().map_err(PolicyError::Malformed)?;
+            if let Some(permitted_subtrees) = nc.permitted_subtrees {
+                constraints
+                    .permitted
+                    .extend(self.build_name_constraints_subtrees(permitted_subtrees));
+            }
+            if let Some(excluded_subtrees) = nc.excluded_subtrees {
+                constraints
+                    .excluded
+                    .extend(self.build_name_constraints_subtrees(excluded_subtrees));
+            }
+        }
+        Ok(())
+    }
+
+    fn apply_name_constraint(
+        &self,
+        constraint: &GeneralName<'work>,
+        san: &GeneralName<'_>,
+    ) -> Result<ApplyNameConstraintStatus, ValidationError> {
+        match (constraint, san) {
+            (GeneralName::DNSName(pattern), GeneralName::DNSName(name)) => {
+                if let Some(pattern) = DNSConstraint::new(pattern.0) {
+                    let name = DNSName::new(name.0).unwrap();
+                    Ok(Applied(pattern.matches(&name)))
+                } else {
+                    Err(PolicyError::Other("malformed DNS name constraint").into())
+                }
+            }
+            (GeneralName::IPAddress(pattern), GeneralName::IPAddress(name)) => {
+                if let Some(pattern) = IPConstraint::from_bytes(pattern) {
+                    let name = IPAddress::from_bytes(name).unwrap();
+                    Ok(Applied(pattern.matches(&name)))
+                } else {
+                    Err(PolicyError::Other("malformed IP name constraint").into())
+                }
+            }
+            _ => Ok(Skipped),
+        }
+    }
+
+    fn apply_name_constraints(
+        &self,
+        constraints: &AccumulatedNameConstraints<'work>,
+        working_cert: &Certificate<'work>,
+    ) -> Result<(), ValidationError> {
+        let extensions = working_cert
+            .extensions()
+            .map_err(|e| ValidationError::Policy(PolicyError::DuplicateExtension(e)))?;
+        if let Some(sans) = extensions.get_extension(&SUBJECT_ALTERNATIVE_NAME_OID) {
+            let sans: SubjectAlternativeName<'_> = sans.value().map_err(PolicyError::Malformed)?;
+            for san in sans.clone() {
```

Comment:
Is it correct that all sans need to be allowed by NCs, or only for the name we're validating?

Reply:
I can't seem to find where that's detailed in the RFC. However, I can confirm that the Go implementation works that way.

Perhaps I can follow up with a Limbo test to make sure that this behaviour is consistent with other implementations.